src/pybind/mgr/dashboard_v2/HACKING.rst

```diff
@@ -508,3 +508,160 @@ The settings management implementation will make sure that if you change a
+  for RBD image creation tasks we could specify ``"rbd/create"`` as the
+  namespace, or conversly ``"rbd/remove"`` for RBD image removal tasks.
+
+* ``metadata`` is a dictionary where we can store key-value pairs that
+  characterize the task. For instance, when creating a task for creating
+  RBD images we can specify the metadata argument as
+  ``{'pool_name': "rbd", image_name': "test-img"}``.
+
+* ``func`` is the python function that implements the operation code, which
+  will be executed asynchronously.
+
+* ``args`` and ``kwargs`` are the positional and named argguments that will be
+  passed to ``func`` when the task manager starts its execution.
+
+The ``TaskManager.run`` method triggers the asynchronous execution of function
+``func`` and returns an ``AsyncTask`` object.
+The ``AsyncTask`` provides the public method ``AsyncTask.wait(timeout)``, which
+can be used to wait for the task to complete up to a timeout defined in seconds
+and provided as an argument. If not argument is provided the ``wait`` method
+blocks until the task is finished.
+
+The ``AsyncTask.wait`` is very useful for tasks that usually are fast
+to execute but that sometimes may take a long time to run.
+The return value of the ``AsyncTask.wait`` method is a pair ``(state, value)``
+where ``state`` is an integer with following possible values:
+
+* ``VALUE_DONE = 0``
+* ``VALUE_EXECUTING = 1``
+* ``VALUE_EXCEPTION = 2``
+
+The ``value`` will store the result of the execution of function ``func`` if
+``state == VALUE_DONE``. If ``state == VALUE_EXECUTING`` then
+``value == None``, and if ``state == VALUE_EXCEPTION`` then ``value`` stores
+the exception object raised by the execution of function ``func``.
+
+The pair ``(namespace, metadata)`` should univocally identify the task being
+run, which means that if you try to trigger a new task that matches the same
+``(namespace, metadata)`` pair of the currently running task, then the new task
+is not created and you get the task object of the current running task.
+
+
+How to get the list of executing and finished asynchronous tasks?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The list of executing and finished tasks is included in the ``Summary``
+controller, which is already polled every 5 seconds by the dashboard frontend.
+But we also provide a dedicated controller to get the same list of executing
+and finished tasks.
+
+The ``Task`` controller exposes the ``/api/task`` endpoint that returns the
+list of executing and finished tasks. This endpoint accepts the ``namespace``
+parameter that accepts a glob expression as its value.
+For instance, an HTTP GET request of the URL ``/api/task?namespace=rbd/*``
+will return all executing and finished tasks which namespace starts with
+``rbd/``.
+
+To prevent the finished tasks list from growing unboundly, the finished tasks
+will be maintained in memory for 1 minute. After a minute, when the finished
```

Comment:
Too short. The absolute minimum for a finished task to stay in the list would be a coffe-break. I'd keep them at least a day.

Reply:
When I thought about how the "finished tasks" list would be used I reached a conclusion that they might not even be necessary, but I decided to keep them anyway and for that reason I choose a very small TTL. For instance, if you trigger a task to delete an RBD image, when the task finishes the image will immediately disappear from the RBD images table, therefore the user will know that the task is finished, without consulting the list of finished tasks.

But I don't mind to change the TTL value to something else as long as we have a quorum of people deciding on the same value.

Reply:
Looks like we have different tasks in mind. "osd/pool/set_pg_count" for example may take quite some time. And as soon as it finishes, it would vanish too soon.

Reply:
As I said before I don't think finished tasks will add much more information to the user, also they are only stored in ceph-mgr memory, which means that the frontend must cope with the fact that a finished task information may not be available.

But I agree that 1 minute might be too small. @LenzGr @ricardoasmarques what value for finished tasks TTL do you think it's appropriate, 1 day?

Reply:
IMO, we should always maintaining in memory the last 20(?) finished tasks, and only use the TTL (1m?) if we have more than 20(?) newer tasks.

This can be useful to list the last finished tasks on the front-end because the end user does not expect to have finished tasks on the list that suddenly disappear, but he can easily understand that the oldest finished tasks are no longer visible.

Reply:
@ricardoasmarques that's a good idea, and easy to implement. @sebastian-philipp what do you think?